src/runtime/virtcontainers/qemu.go

```diff
@@ -1020,47 +1053,199 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
 }
 
 func (q *qemu) qmpSetup() error {
-	q.qmpMonitorCh.Lock()
-	defer q.qmpMonitorCh.Unlock()
+	return q.startQMPInstance(&q.qmpMonitorCh, "monitor-qmp", true)
+}
 
-	if q.qmpMonitorCh.qmp != nil {
+func (q *qemu) startQMPInstance(qmpChan *qmpChannel, subsystem string, enableEvents bool) error {
+	qmpChan.Lock()
+	defer qmpChan.Unlock()
+
+	if qmpChan.qmp != nil {
 		return nil
 	}
 
-	cfg := govmmQemu.QMPConfig{Logger: newQMPLogger()}
+	var events chan govmmQemu.QMPEvent
+
+	if enableEvents {
+		events = make(chan govmmQemu.QMPEvent)
+		go q.loopQMPEvent(events)
+	}
+
+	// It's safe to pass a nil EventCh
+	cfg := govmmQemu.QMPConfig{
+		Logger: qmpLogger{
+			logger: virtLog.WithField("subsystem", subsystem),
+		},
+		EventCh: events,
+	}
 
 	// Auto-closed by QMPStart().
 	disconnectCh := make(chan struct{})
 
-	qmp, _, err := govmmQemu.QMPStart(q.qmpMonitorCh.ctx, q.qmpMonitorCh.path, cfg, disconnectCh)
+	qmp, _, err := govmmQemu.QMPStart(qmpChan.ctx, qmpChan.path, cfg, disconnectCh)
 	if err != nil {
 		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
 		return err
 	}
 
-	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	err = qmp.ExecuteQMPCapabilities(qmpChan.ctx)
 	if err != nil {
 		qmp.Shutdown()
 		q.Logger().WithError(err).Error(qmpCapErrMsg)
 		return err
 	}
-	q.qmpMonitorCh.qmp = qmp
-	q.qmpMonitorCh.disconn = disconnectCh
+	qmpChan.qmp = qmp
+	qmpChan.disconn = disconnectCh
+
+	return nil
+}
+
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
+	for {
+		select {
+		case e, open := <-event:
+			if !open {
+				q.Logger().Infof("QMP event channel closed")
+				return
+			}
+			q.Logger().Debugf("got new QMP event %+v", e)
+			if e.Name == "GUEST_PANICKED" {
+				go q.handleGuestPanic()
+			}
+		}
+	}
+}
+
+func (q *qemu) handleGuestPanic() {
+	if q.config.GuestMemoryDumpPath != "" {
```

Comment:
I'd move this test into `q.dumpGuestMemory()` (or also add it to that function).